src/com/xilinx/rapidwright/rwroute/RoutableNode.java

```diff
@@ -0,0 +1,570 @@
+/*
+ * 
+ * Copyright (c) 2021 Ghent University. 
+ * All rights reserved.
+ *
+ * Author: Yun Zhou, Ghent University.
+ *
+ * This file is part of RapidWright. 
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * 
+ */
+
+package com.xilinx.rapidwright.rwroute;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.xilinx.rapidwright.design.SitePinInst;
+import com.xilinx.rapidwright.device.IntentCode;
+import com.xilinx.rapidwright.device.Node;
+import com.xilinx.rapidwright.device.Tile;
+import com.xilinx.rapidwright.device.TileTypeEnum;
+import com.xilinx.rapidwright.device.Wire;
+import com.xilinx.rapidwright.router.RouteThruHelper;
+import com.xilinx.rapidwright.timing.delayestimator.DelayEstimatorBase;
+
+/**
+ * A RoutableNode Object, denoted as rnode, is a vertex of the routing resource graph.
+ * It implements {@link Routable} and is created based on a {@link Node} Object.
+ */
+public class RoutableNode implements Routable{
+	/** A unique index of a rnode */
+	private int index;
+	/** The associated {@link Node} instance */
+	private Node node;
+	/** The type of a rnode*/
+	private RoutableType type;
+	/** The tileXCoordinate and tileYCoordinate of the INT tile that the associated node stops at */
+	private short endTileXCoordinate;
+	private short endTileYCoordinate;
```

Comment:
Is considering just the coordinates of the end tile sufficient in all cases? It seems that this is used for a lot of things, including for finding the geometric center, bounding box, etc.

One extension I've been looking to add is to constrain the routing within a `PBlock`, and this requires knowing the {min,max}{X,Y} of the node.

Reply:
Knowing the end tile of nodes should be enough. For example, for a horizontal node going from left to right, its min X (the starting point) is the end tile X coordinate of its uphill nodes.